### ISSUE TITLE
"systemctl preset-all" is broken? (when triggered by missing /etc/machine-id during boot)

### ISSUE THREAD

**Issue by 057a3dd61f99517a3afea0051a49cb27994f94d** [systemd/systemd#10253]
**systemd version the issue has been seen with**
> systemd-239

**Used distribution**
 > custom

**Expected behaviour you didn't see**
 > …

**Unexpected behaviour you saw**
 > Creates meaningless dependency `machines.target.wants/systemd-nspawn@.service` and
`getty.target.wants/serial-getty@.service` under `/etc/systemd/system`,
pointing to their template units under `/lib/systemd/system` respectively but
without parameter.

**Steps to reproduce the problem**
Run `systemd` with missing `/etc/machine-id`.

I'm running `systemd-239` with `linux-v4.4.156` under `qemu-3.0.0`, the problem I
encountered is the same as #8755, I also found out that the exact step
to trigger this buggy behavior is to run `systemd` with `/etc/machine-id`
missing during the first boot, everything is working fine if you do
`touch /etc/machine-id` and run
`systemd` with `/etc/machine-id` empty. Other than the two unwanted dependency
there're also some other differences with missing `/etc/machine-id`:

- `systemd` asks for locale name and message locale name.
- the journal says `Running with unpopulated /etc`.
- the journal says `Starting First Boot Wizard`.

All of the above will not be seen when running with empty `/etc/machine-id`,
when I read `machine-id(5)` it says `systemd` will use a randomly generated
UUID if `/etc/machine-id` is empty or missing, it never says it's an error
to run `systemd` with `/etc/machine-id` missing, is this a bug?

**Comment by poettering** [comment]
can you provide full debug logs of the early boot phase please?

Are you using an initrd? Is your root disk mounted writable by your kernel? or by the initrd? or by the host system? or not at all?

**Comment by poettering** [comment]
(normally we should simply refuse to boot if we can't set up the machine ID)

**Comment by poettering** [comment]
That log says: "Oct 02 23:14:37 somebody systemd[1]: Initializing machine ID from random generator."

so all appears good with the machine id?

Not following what the issue is supposed to be?

**Comment by 057a3dd61f99517a3afea0051a49cb27994f94d** [comment]
The problem is that it will create two meaningless dependency files:

- `/etc/systemd/system/machines.target.wants/systemd-nspawn@.service`, which points
to `/lib/systemd/system/systemd-nspawn@.service`.
- `/etc/systemd/system/getty.target.wants/serial-getty@.service`, which points to
`/lib/systemd/system/serial-getty@.service`.

Aren't those files pointless? They point to template units while they don't provide
parameters themselves, and those services will fail in all the subsequent boots.
While the problem doesn't exist at all if I run `systemd` with `/etc/machine-id`
empty instead of missing during the first boot.

**Comment by 057a3dd61f99517a3afea0051a49cb27994f94d** [comment]
Some more details.

When booting with empty `/etc/machine-id`, nothing is modified under
`/etc/systemd/systemd`, however,
here's a list of services `systemd` added automatically when booting with
missing `/etc/machine-id`:

```
/etc/systemd/system/ctrl-alt-del.target => /lib/systemd/system/poweroff.target
/etc/systemd/system/dbus-org.freedesktop.timesync1.service => /lib/systemd/system/systemd-timesyncd.service
/etc/systemd/system/getty.target.wants/
/etc/systemd/system/getty.target.wants/console-getty.service => /lib/systemd/system/console-getty.service
/etc/systemd/system/getty.target.wants/serial-getty@.service => /lib/systemd/system/serial-getty@.service
/etc/systemd/system/machines.target.wants/
/etc/systemd/system/machines.target.wants/systemd-nspawn@.service => /lib/systemd/system/systemd-nspawn@.service
/etc/systemd/system/multi-user.target.wants/
/etc/systemd/system/multi-user.target.wants/systemd-journal-upload.service => /lib/systemd/system/systemd-journal-upload.service
/etc/systemd/system/sysinit.target.wants/
/etc/systemd/system/sysinit.target.wants/debug-shell.service => /lib/systemd/system/debug-shell.service
/etc/systemd/system/sysinit.target.wants/systemd-time-wait-sync.service => /lib/systemd/system/systemd-time-wait-sync.service
```

And these two installed services `serial-getty@.service` `systemd-nspawn@.service`
don't make any sense. To summarize, when booting with missing `/etc/machine-id`,
the problems are:

- Two meaningless services that will always fail are installed automatically.
- Lots of additional services are installed, comparing to booting with empty
`/etc/machine-id` where no additional service is installed, the behaviors
are so different and it's really a big surprise.

Since the journal says `Populated /etc with preset unit settings.` with missing
`/etc/machine-id`, perhaps the problem lies within how `systemd` actually
did that.

**Comment by 057a3dd61f99517a3afea0051a49cb27994f94d** [comment]
Yes, I believe `systemctl preset-all` is indeed problematic, I'm using
`systemd-239` without any patch or special `.preset` files, and I tried
to run `systemctl preset-all` after a fresh install of `systemd-239`
inside a container instead of booting with `qemu`, here's the output:

```
Created symlink /etc/systemd/system/dbus-org.freedesktop.timesync1.service → /lib/systemd/system/systemd-timesyncd.service.
Created symlink /etc/systemd/system/ctrl-alt-del.target → /lib/systemd/system/reboot.target.
Created symlink /etc/systemd/system/machines.target.wants/systemd-nspawn@.service → /lib/systemd/system/systemd-nspawn@.service.
Created symlink /etc/systemd/system/getty.target.wants/serial-getty@.service → /lib/systemd/system/serial-getty@.service.
Created symlink /etc/systemd/system/sysinit.target.wants/systemd-time-wait-sync.service → /lib/systemd/system/systemd-time-wait-sync.service.
```

I don't think those two services `systemd-nspawn@.service` and
`serial-getty@.service` make sense, and they always fail.

**Comment by poettering** [comment]
Fix waiting in #10954
